File: lesson-3/composition2/Light.js
"use strict";
class Light
{
    #sm = new LightSm();

    #smInterfaceObj = {
        count: 0,
        lightBlue: this.#lightBlue.bind(this),
        lightYellow: this.#lightYellow.bind(this),
        lightRed: this.#lightRed.bind(this),
        print: this.#print.bind(this)
    };

    constructor() {
        // give state machine a reference to the interface object
        this.#sm.vars.interfaceObj = this.#smInterfaceObj;
    }

    //-------------- public methods ----------------

    start() {
        this.#sm.start();
    }

    inc() {
        this.#dispatchEvent(LightSm.EventId.INC);
    }

    dim() {
        this.#dispatchEvent(LightSm.EventId.DIM);
    }

    off() {
        this.#dispatchEvent(LightSm.EventId.OFF);
    }

    //-------------- private methods ----------------

    #dispatchEvent(eventId) {
        this.#sm.dispatchEvent(eventId);
        document.getElementById("count-var").innerText = this.#smInterfaceObj.count + "";
    }

    #lightBlue() {
        this.#print("BLUE ");
    }

    #lightYellow() {
        this.#print("YELLOW");
    }

    #lightRed() {
        this.#print("RED");
    }

    #print(message) {
        const output = document.getElementById("output");
        output.append(message + "\n");
    }
}

File: lesson-4/c99-gcc/Light.c
#include "Light.h"
#include <stdio.h>

// ANSI escape codes for colors
#define RESET   "\033[0m"
#define BLUE    "\033[0;34m"
#define YELLOW  "\033[0;33m"
#define RED     "\033[0;31m"

void Light_blue(void)
{
    printf(BLUE "Light is blue\n" RESET);
}

void Light_yellow(void)
{
    printf(YELLOW "Light is yellow\n" RESET);
}

void Light_red(void)
{
    printf(RED "Light is red\n" RESET);
}
